react/features/chat/components/native/Chat.js

```diff
@@ -0,0 +1,168 @@
+            modalStyle.push(styles.solidModalBackdrop);
+        }
+
+        return (
+            <Modal
+                onRequestClose = { this.props._onToggleChat }
+                visible = { this.props._isOpen }>
+                <Header>
+                    <BackButton onPress = { this.props._onToggleChat } />
+                    <HeaderLabel labelKey = 'chat.title' />
+                </Header>
+                <SafeAreaView style = { modalStyle }>
+                    <GiftedChat
+                        messages = { messages }
+                        onSend = { this._onSend }
+                        renderMessage = { this._renderMessage } />
+                </SafeAreaView>
+            </Modal>
+        );
+    }
+
+    _onSend: (Array<Object>) => void;
+
+    /**
+     * Callback to trigger a message send action.
+     *
+     * @param {string} message - The chat message to display.
+     * @returns {void}
+     */
+    _onSend([ message ]) {
+        this.props._onSendMessage(message.text);
+    }
+
+    _renderMessage: Object => React$Element<*>
+
+    /**
+     * Renders a single message.
+     *
+     * @param {Object} messageProps - The message props object to be rendered.
+     * @returns {React$Element<*>}
+     */
+    _renderMessage(messageProps) {
+        const { currentMessage } = messageProps;
+
+        return (
+            <ChatMessage message = { currentMessage } />
+        );
+    }
+
+    _transformMessage: (Object, number) => Object;
+
+    /**
+     * Transforms a Jitsi message object to a format that gifted-chat can
+     * handle.
+     *
+     * @param {Object} message - The chat message in our internal format.
+     * @param {number} index - The index of the message in the array.
+     * @returns {Object}
+     */
+    _transformMessage(message, index) {
+        const system = message.messageType === 'error';
+
+        return (
+            {
+                _id: index,
+                createdAt: new Date(message.timestamp),
+                messageType: message.messageType,
+                system,
+                text: system
+                    ? this.props.t('chat.error', {
+                        error: message.error,
+                        originalText: message.message
+                    })
+                    : message.message,
+                user: {
+                    _id: message.id,
+                    name: message.displayName
+                }
+            }
+        );
+    }
+}
+
+/**
+ * Maps part of the Redux state to the props of this component.
+ *
+ * @param {Object} state - The Redux state.
+ * @returns {{
+ *     _solidBackground: boolean
+ * }}
+ */
+function _mapStateToProps(state) {
+    const abstractReduxProps = _abstractMapStateToProps(state);
+
+    return {
+        ...abstractReduxProps,
+
+        // Gifted chat requires the messages to be reverse ordered.
+        _messages: [
+            ...abstractReduxProps._messages
+        ].reverse(),
```

Comment:
Thing about this more, does creating a new array here cause more re-renders than necessary because a new array is being made each time mapStateToProps is called?

Reply:
True, I guess this should go back to render then